grafana/grafana
# Loki: Fix incorrect evaluation of real and extracted labels in context

In this PR we are changing a logic to determine which labels are actual and which are extracted. Previously, we compared `label names` from selected log row to all `labelKeys` for that Loki instance. This works correctly if query does not have parser, but if query has parser, then in edge cases, `extractedLabel` can have the same name as `actualLabel` from different log stream.

To fix this, we are:

1. not changing anything in a case when query has no parser
2. fetching labels for selected stream if query has parser


In `fetchSeriesLabels` I have removed code bellow that was preventing from running fetchSeriesLabels if request with the same cacheKey has already started, but we haven't received values yet. The issue here is that implementation is actually not a correct implementation because it falsely returns empty object and there is no way to know if the value is actually empty object, or if we are canceling the request on purpose. If we want to improve fetchSeriesLabels caching, we need to change and improve the logic as whole, because incorrectly returning empty labels can be missleading.

```
      // Clear value when requesting new one. Empty object being truthy also makes sure we don't request twice.
      this.seriesCache.set(cacheKey, {});
```

Fixes: https://github.com/grafana/grafana/issues/47937

How to test:
1. Manually change data.js and replace following line [](https://github.com/grafana/grafana/blob/main/devenv/docker/blocks/loki/data/data.js#L178) `await lokiSendLogLine(timestampNs, JSON.stringify(item), {place:'moon', ...sharedLabels});` => `await lokiSendLogLine(timestampNs, JSON.stringify(item), {place:'moon', ...sharedLabels, level: 'info'});` so we create `actual` label with the same name as `extracted` label, but only in `place:moon` stream
2. Run `make devenv sources=loki`
3. Open Explore and run `{place="luna"} | logfmt` query
4. Open log conext and you should not see `level` in the list of actual labels. It should be available only in `extracted` labels listed

Fixed functionality (this branch):

https://user-images.githubusercontent.com/30407135/234016689-32249a94-05ac-42e2-b2c5-316d0b480a63.mov


Broken functionality (main):

https://user-images.githubusercontent.com/30407135/234016198-c45a790b-6c4a-487f-bf94-f5b1faa0c225.mov
